File: static/chess_app/legalMoves.js
import Piece from './piece.js'


class Move {
    constructor(pieceType, oldPosition, newPosition) {
        this.piece = pieceType;
        this.oldSquare = oldPosition;
        this.newSquare = newPosition;
    }
}


export default function getLegalMoves(player, piece, piecePosition, positions, previousMove, removeChecks) {
    if (player !== piece.color) {
        return [];
    }

    if (piece.type === 'pawn') {
        return getLegalPawnMoves(player, piecePosition, positions, removeChecks, previousMove);
    }

    if (piece.type === 'knight') {
        return getLegalKnightMoves(player, piecePosition, positions, removeChecks);
    }

    if (piece.type === 'bishop') {
        return getLegalBishopMoves(player, piecePosition, positions, removeChecks);
    }

    if (piece.type === 'rook') {
        return getLegalRookMoves(player, piecePosition, positions, removeChecks);
    }

    if (piece.type === 'queen') {
        return getLegalQueenMoves(player, piecePosition, positions, removeChecks);
    }

    if (piece.type === 'king') {
        return getLegalKingMoves(player, piecePosition, positions, removeChecks);
    }
}


function getLegalPawnMoves(player, piecePosition, positions, removeChecks, previousMove) {
    const rank = getRankNumber(player, piecePosition);
    const file = getFileNumber(player, piecePosition);
    let legalPawnMoves = [];

    const oneForward = getSquareCode(player, file, rank+1);
    if (squareIsFree(positions, oneForward)) {
        legalPawnMoves.push(new Move('pawn', piecePosition, oneForward));
    }

    const twoForward = getSquareCode(player, file, rank+2);
    if (rank === 2 && squareIsFree(positions, oneForward) && squareIsFree(positions, twoForward)) {
        legalPawnMoves.push(new Move('pawn', piecePosition, twoForward));
    }

    const takeLeft = getSquareCode(player, file-1, rank+1);
    if (opponentOccupiesSquare(player, positions, takeLeft)) {
        legalPawnMoves.push(new Move('pawn', piecePosition, takeLeft));
    }

    const takeRight = getSquareCode(player, file+1, rank+1);
    if (opponentOccupiesSquare(player, positions, takeRight)) {
        legalPawnMoves.push(new Move('pawn', piecePosition, takeRight));
    }

    if (removeChecks) {
        return removeCheckingMoves(legalPawnMoves, positions, player);
    } else {
        return legalPawnMoves;
    }
}


function getLegalKnightMoves(player, piecePosition, positions, removeChecks) {
    const rank = getRankNumber(player, piecePosition);
    const file = getFileNumber(player, piecePosition);
    let legalKnightMoves = [];

    const forwardLeft = getSquareCode(player, file-1, rank+2);
    legalKnightMoves = evaluateKnightMove(new Move('knight', piecePosition, forwardLeft), player, positions, legalKnightMoves);

    const forwardRight = getSquareCode(player, file+1, rank+2);
    legalKnightMoves = evaluateKnightMove(new Move('knight', piecePosition, forwardRight), player, positions, legalKnightMoves);

    const leftForward = getSquareCode(player, file-2, rank+1);
    legalKnightMoves = evaluateKnightMove(new Move('knight', piecePosition, leftForward), player, positions, legalKnightMoves);

    const rightForward = getSquareCode(player, file+2, rank+1);
    legalKnightMoves = evaluateKnightMove(new Move('knight', piecePosition, rightForward), player, positions, legalKnightMoves);

    const backwardLeft = getSquareCode(player, file-1, rank-2);
    legalKnightMoves = evaluateKnightMove(new Move('knight', piecePosition, backwardLeft), player, positions, legalKnightMoves);

    const backwardRight = getSquareCode(player, file+1, rank-2);
    legalKnightMoves = evaluateKnightMove(new Move('knight', piecePosition, backwardRight), player, positions, legalKnightMoves);

    const leftBackward = getSquareCode(player, file-2, rank-1);
    legalKnightMoves = evaluateKnightMove(new Move('knight', piecePosition, leftBackward), player, positions, legalKnightMoves);

    const rightBackward = getSquareCode(player, file+2, rank-1);
    legalKnightMoves = evaluateKnightMove(new Move('knight', piecePosition, rightBackward), player, positions, legalKnightMoves);

    if (removeChecks) {
        return removeCheckingMoves(legalKnightMoves, positions, player);
    } else {
        return legalKnightMoves;
    }
}


function evaluateKnightMove(move, player, positions, legalKnightMoves) {
    if (move.newSquare !== null && !iOccupySquare(player, positions, move.newSquare)) {
        legalKnightMoves.push(move);
    }
    return legalKnightMoves;
}


function getLegalBishopMoves(player, piecePosition, positions, removeChecks) {
    const rank = getRankNumber(player, piecePosition);
    const file = getFileNumber(player, piecePosition);
    let legalBishopMoves = [];

    const maxLeftUpDistance = Math.min(file, (9-rank));
    for (let i = 1; i < maxLeftUpDistance; i++) {
        const leftUp = getSquareCode(player, file-i, rank+i);
        if (leftUp !== null && !iOccupySquare(player, positions, leftUp)) {
            legalBishopMoves.push(new Move('bishop', piecePosition, leftUp));
            if (!squareIsFree(positions, leftUp)) {
                break;
            }
        } else {
            break;
        }
    }

    const maxRightUpDistance = Math.min((9-file), (9-rank));
    for (let i = 1; i < maxRightUpDistance; i++) {
        const rightUp = getSquareCode(player, file+i, rank+i);
        if (rightUp !== null && !iOccupySquare(player, positions, rightUp)) {
            legalBishopMoves.push(new Move('bishop', piecePosition, rightUp));
            if (!squareIsFree(positions, rightUp)) {
                break;
            }
        } else {
            break;
        }
    }

    const maxLeftDownDistance = Math.min(file, rank);
    for (let i = 1; i < maxLeftDownDistance; i++) {
        const leftDown = getSquareCode(player, file-i, rank-i);
        if (leftDown !== null && !iOccupySquare(player, positions, leftDown)) {
            legalBishopMoves.push(new Move('bishop', piecePosition, leftDown));
            if (!squareIsFree(positions, leftDown)) {
                break;
            }
        } else {
            break;
        }
    }

    const maxRightDownDistance = Math.min((9-file), rank);
    for (let i = 1; i < maxRightDownDistance; i++) {
        const rightDown = getSquareCode(player, file+i, rank-i);
        if (rightDown !== null && !iOccupySquare(player, positions, rightDown)) {
            legalBishopMoves.push(new Move('bishop', piecePosition, rightDown));
            if (!squareIsFree(positions, rightDown)) {
                break;
            }
        } else {
            break;
        }
    }

    if (removeChecks) {
        return removeCheckingMoves(legalBishopMoves, positions, player);
    } else {
        return legalBishopMoves;
    }
}


function getLegalRookMoves(player, piecePosition, positions, removeChecks) {
    const rank = getRankNumber(player, piecePosition);
    const file = getFileNumber(player, piecePosition);
    let legalRookMoves = [];

    for (let i = 1; i < (9-rank); i++) {
        const up = getSquareCode(player, file, rank+i);
        if (up !== null && !iOccupySquare(player, positions, up)) {
            legalRookMoves.push(new Move('rook', piecePosition, up));
            if (!squareIsFree(positions, up)) {
                break;
            }
        } else {
            break;
        }
    }

    for (let i = 1; i < (9-file); i++) {
        const right = getSquareCode(player, file+i, rank);
        if (right !== null && !iOccupySquare(player, positions, right)) {
            legalRookMoves.push(new Move('rook', piecePosition, right));
            if (!squareIsFree(positions, right)) {
                break;
            }
        } else {
            break;
        }
    }

    for (let i = 1; i < rank; i++) {
        const down = getSquareCode(player, file, rank-i);
        if (down !== null && !iOccupySquare(player, positions, down)) {
            legalRookMoves.push(new Move('rook', piecePosition, down));
            if (!squareIsFree(positions, down)) {
                break;
            }
        } else {
            break;
        }
    }

    for (let i = 1; i < file; i++) {
        const left = getSquareCode(player, file-i, rank);
        if (left !== null && !iOccupySquare(player, positions, left)) {
            legalRookMoves.push(new Move('rook', piecePosition, left));
            if (!squareIsFree(positions, left)) {
                break;
            }
        } else {
            break;
        }
    }

    if (removeChecks) {
        return removeCheckingMoves(legalRookMoves, positions, player);
    } else {
        return legalRookMoves;
    }
}


function getLegalQueenMoves(player, piecePosition, positions, removeChecks) {
    const legalBishopMoves = getLegalBishopMoves(player, piecePosition, positions);
    const legalRookMoves = getLegalRookMoves(player, piecePosition, positions);
    let legalQueenMoves = legalBishopMoves.concat(legalRookMoves);

    if (removeChecks) {
        return removeCheckingMoves(legalQueenMoves, positions, player);
    } else {
        return legalQueenMoves;
    }
}


function getLegalKingMoves(player, piecePosition, positions, removeChecks) {
    const rank = getRankNumber(player, piecePosition);
    const file = getFileNumber(player, piecePosition);

    let legalKingMoves = [];

    const up = new Move('king', piecePosition, getSquareCode(player, file, rank+1));
    legalKingMoves = evaluateKingMove(up, player, positions, legalKingMoves);

    const rightUp = new Move('king', piecePosition, getSquareCode(player, file+1, rank+1));
    legalKingMoves = evaluateKingMove(rightUp, player, positions, legalKingMoves);

    const right = new Move('king', piecePosition, getSquareCode(player, file+1, rank));
    legalKingMoves = evaluateKingMove(right, player, positions, legalKingMoves);

    const rightDown = new Move('king', piecePosition, getSquareCode(player, file+1, rank-1));
    legalKingMoves = evaluateKingMove(rightDown, player, positions, legalKingMoves);

    const down = new Move('king', piecePosition, getSquareCode(player, file, rank-1));
    legalKingMoves = evaluateKingMove(down, player, positions, legalKingMoves);

    const leftDown = new Move('king', piecePosition, getSquareCode(player, file-1, rank-1));
    legalKingMoves = evaluateKingMove(leftDown, player, positions, legalKingMoves);

    const left = new Move('king', piecePosition, getSquareCode(player, file-1, rank));
    legalKingMoves = evaluateKingMove(left, player, positions, legalKingMoves);

    const leftUp = new Move('king', piecePosition, getSquareCode(player, file-1, rank+1));
    legalKingMoves = evaluateKingMove(leftUp, player, positions, legalKingMoves);

    if (removeChecks) {
        return removeCheckingMoves(legalKingMoves, positions, player);
    } else {
        return legalKingMoves;
    }
}


function evaluateKingMove(move, player, positions, legalKingMoves) {
    if (move.newSquare !== null && !iOccupySquare(player, positions, move.newSquare)) {
        legalKingMoves.push(move);
    }
    return legalKingMoves;
}


function getOtherPlayer(player) {
    if (player === 'white') {
        return 'black';
    } else if (player === 'black') {
        return 'white';
    }
}


function squareIsFree(positions, squareCode) {
    return !positions.hasOwnProperty(squareCode);
}


function opponentOccupiesSquare(player, positions, squareCode) {
    if (squareIsFree(positions, squareCode)) {
        return false;
    } else {
        return !(positions[squareCode]['color'] === player);
    }
}


function iOccupySquare(player, positions, squareCode) {
    if (squareIsFree(positions, squareCode)) {
        return false;
    } else {
        return (positions[squareCode]['color'] === player);
    }
}


function getFileNumber(player, squareCode) {
    if (player === 'white') {
        return squareCode.charCodeAt(0) - 'a'.charCodeAt(0) + 1;
    } else if (player === 'black') {
        return 9 - (squareCode.charCodeAt(0) - 'a'.charCodeAt(0) + 1);
    }
}


function getRankNumber(player, squareCode) {
    if (player === 'white') {
        return Number(squareCode[1]);
    } else if (player === 'black') {
        return 9 - Number(squareCode[1]);
    }
}


function getSquareCode(player, file, rank) {
    if (file <= 0 || file > 8 || rank <= 0 || rank > 8) {
        return null;
    }
    if (player === 'white') {
        const fileCode = String.fromCharCode(file - 1 + 'a'.charCodeAt(0));
        const rankCode = rank.toString();
        return fileCode + rankCode;
    } else if (player === 'black') {
        const fileCode = String.fromCharCode(8 - file + 'a'.charCodeAt(0));
        const rankCode = (9 - rank).toString();
        return fileCode + rankCode;
    }
}   


function removeCheckingMoves(legalMoves, positions, player) {
    let legalMovesWithoutChecks = [];

    legalMoves.forEach(move => {
        if (!willBeCheck(move, positions, player)) {
            legalMovesWithoutChecks.push(move);
        }
    })

    return legalMovesWithoutChecks
}


function willBeCheck(move, positions, player) {
    const positionsCopy = {...positions};
    delete positionsCopy[move.oldSquare];
    positionsCopy[move.newSquare] = new Piece(player, move.piece);
    return isCheck(positionsCopy, player);
}


function isCheck(positions, player) {
    const playersKingPosition = Object.keys(positions).find(field => positions[field]['color'] === player && positions[field]['type'] === 'king')
    for (const occupiedField in positions) {
        const currentPiece = positions[occupiedField];
        if (currentPiece.color != player) {
            const legalMoves = getLegalMoves(getOtherPlayer(player), currentPiece, occupiedField, positions, null, false);
            if (legalMoves.find(move => move.newSquare === playersKingPosition)) {
                return true;
            }
        }
    }
    return false;
}

File: src/legal_moves.py
import copy
from chess_app.src.move import Move


class LegalMoves:

    def __init__(self, player, chessboard_state, piece=None, remove_checks=True):
        self.player = player
        self.chessboard_state = chessboard_state
        self.piece = piece
        self.should_remove_checks = remove_checks

        if self.piece is not None:
            self.rank_number = self.get_rank_number(self.piece['position'])
            self.file_number = self.get_file_number(self.piece['position'])
        else:
            self.rank_number = None
            self.file_number = None

        self.legal_moves = []

    def set_piece(self, piece):
        self.piece = piece

        self.rank_number = self.get_rank_number(piece['position'])
        self.file_number = self.get_file_number(piece['position'])

    def get_all_legal_moves(self):
        all_legal_moves = []
        for square_id, square_data in self.chessboard_state.items():
            if square_data['piece'] is not None and square_data['piece']['color'] == self.player:
                self.set_piece(square_data['piece'])
                all_legal_moves.extend(self.get_legal_moves())

        return all_legal_moves

    def get_legal_moves(self):
        self.legal_moves = []
        if self.player != self.piece['color']:
            return []
        
        if self.piece['type'] == 'pawn':
            self.add_legal_pawn_moves()
        if self.piece['type'] == 'knight':
            self.add_legal_knight_moves()
        if self.piece['type'] == 'bishop':
            self.add_legal_bishop_moves()
        if self.piece['type'] == 'rook':
            self.add_legal_rook_moves()
        if self.piece['type'] == 'queen':
            self.add_legal_queen_moves()
        if self.piece['type'] == 'king':
            self.add_legal_king_moves()

        if self.should_remove_checks:
            self.remove_checks()

        return self.legal_moves

    def add_legal_pawn_moves(self):
        one_forward = self.get_square_code(self.file_number, self.rank_number+1)
        if one_forward:
            self.assess_move(one_forward, [self.square_is_free(one_forward)])

        two_forward = self.get_square_code(self.file_number, self.rank_number+2)
        if two_forward:
            self.assess_move(two_forward, [self.square_is_free(two_forward), self.square_is_free(one_forward), self.rank_number==2])

        take_left = self.get_square_code(self.file_number-1, self.rank_number+1)
        if take_left:
            self.assess_move(take_left, [self.opponent_occupies_square(take_left)])

        take_right = self.get_square_code(self.file_number+1, self.rank_number+1)
        if take_right:
            self.assess_move(take_right, [self.opponent_occupies_square(take_right)])

    def add_legal_knight_moves(self):
        possible_moves = [(self.file_number+1, self.rank_number-2), (self.file_number+2, self.rank_number-1),
                          (self.file_number+1, self.rank_number+2), (self.file_number+2, self.rank_number+1),
                          (self.file_number-1, self.rank_number-2), (self.file_number-2, self.rank_number-1),
                          (self.file_number-1, self.rank_number+2), (self.file_number-2, self.rank_number+1),]

        for possible_move in possible_moves:
            new_position = self.get_square_code(possible_move[0], possible_move[1])
            if new_position:
                self.assess_move(new_position, [not self.i_occupy_square(new_position)])

    def add_legal_bishop_moves(self):
        possible_directions = [[(self.file_number-i, self.rank_number+i) for i in range(1, min(self.file_number, (9-self.rank_number)))],
                               [(self.file_number+i, self.rank_number+i) for i in range(1, min((9-self.file_number), (9-self.rank_number)))],
                               [(self.file_number-i, self.rank_number-i) for i in range(1, min(self.file_number, self.rank_number))],
                               [(self.file_number+i, self.rank_number-i) for i in range(1, min((9-self.file_number), self.rank_number))]]

        for possible_moves in possible_directions:
            for possible_move in possible_moves:
                new_position = self.get_square_code(possible_move[0], possible_move[1])
                if new_position:
                    path_blocked = self.assess_move(new_position, [not self.i_occupy_square(new_position)])
                    if path_blocked:
                        break

    def add_legal_rook_moves(self):
        possible_directions = [[(self.file_number, self.rank_number+i) for i in range(1, (9-self.rank_number))],
                               [(self.file_number+i, self.rank_number) for i in range(1, (9-self.file_number))],
                               [(self.file_number, self.rank_number-i) for i in range(1, self.rank_number)],
                               [(self.file_number-i, self.rank_number) for i in range(1, self.file_number)]]

        for possible_moves in possible_directions:
            for possible_move in possible_moves:
                new_position = self.get_square_code(possible_move[0], possible_move[1])
                if new_position:
                    path_blocked = self.assess_move(new_position, [not self.i_occupy_square(new_position)])
                    if path_blocked:
                        break

    def add_legal_queen_moves(self):
        self.add_legal_bishop_moves()
        self.add_legal_rook_moves()

    def add_legal_king_moves(self):
        possible_moves = [(self.file_number, self.rank_number+1), (self.file_number+1, self.rank_number+1),
                          (self.file_number+1, self.rank_number), (self.file_number+1, self.rank_number-1),
                          (self.file_number, self.rank_number-1), (self.file_number-1, self.rank_number-1),
                          (self.file_number-1, self.rank_number), (self.file_number-1, self.rank_number+1)]

        for possible_move in possible_moves:
            new_position = self.get_square_code(possible_move[0], possible_move[1])
            if new_position:
                self.assess_move(new_position, [not self.i_occupy_square(new_position)])

    def assess_move(self, new_position, conditions):
        if all(conditions):
            new_move = Move(self.piece['type'], self.piece['position'], new_position)
            self.legal_moves.append(new_move)
            if not self.square_is_free(new_position):
                return True
            return False
        return True

    def remove_checks(self):
        moves_without_checks = []

        for legal_move in self.legal_moves:
            next_chessboard_state = copy.deepcopy(self.chessboard_state)
            next_chessboard_state[legal_move.new]['piece'] = copy.copy(self.chessboard_state[legal_move.old]['piece'])
            next_chessboard_state[legal_move.old]['piece'] = None

            if not self.can_next_player_take_king(next_chessboard_state):
                moves_without_checks.append(legal_move)
        
        self.legal_moves = moves_without_checks            

    def can_next_player_take_king(self, next_chessboard_state):
        next_players_legal_moves = LegalMoves(self.get_other_player(), next_chessboard_state, remove_checks=False).get_all_legal_moves()
        for move in next_players_legal_moves:
            new_square = next_chessboard_state[move.new]
            if new_square['piece'] and new_square['piece']['color'] == self.player and new_square['piece']['type'] == 'king':
                return True
        return False

    def square_is_free(self, square_code):
        return self.chessboard_state[square_code]['piece'] == None

    def opponent_occupies_square(self, square_code):
        if self.square_is_free(square_code):
            return False
        else:
            return self.chessboard_state[square_code]['piece']['color'] != self.player

    def i_occupy_square(self, square_code):
        if self.square_is_free(square_code):
            return False
        else:
            return self.chessboard_state[square_code]['piece']['color'] == self.player

    def get_file_number(self, square_code):
        if self.player == 'white':
            return ord(square_code[0]) - ord('a') + 1
        elif self.player == 'black':
            return 9 - (ord(square_code[0]) - ord('a') + 1)

    def get_rank_number(self, square_code):
        if self.player == 'white':
            return int(square_code[1])
        elif self.player == 'black':
            return 9 - int(square_code[1])

    def get_square_code(self, file_number, rank_number):
        if file_number <= 0 or file_number > 8 or rank_number <= 0 or rank_number > 8:
            return None
        if self.player == 'white':
            file_code = chr(file_number - 1 + ord('a'))
            rank_code = str(rank_number)
            return file_code + rank_code
        elif self.player == 'black':
            file_code = chr((9 - file_number) - 1 + ord('a'))
            rank_code = str(9 - rank_number)
            return file_code + rank_code

    def get_other_player(self):
        if self.player == 'white':
            return 'black'
        else:
            return 'white'
